### PR TITLE
fix unsigned flag for `COM_STMT_EXECUTE` when `new_params_bind_flag` is set

### PR DESCRIPTION
**Pull request by JCOR11599**
In the previous implementation, we assumed that the way the MySQL Protocol specifies [Column Definitions](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_com_query_response_text_resultset_column_definition.html) is the same as how it specifies parameter types for `COM_STMT_EXECUTE`. The difference lies specifically in the flags that come after the field type.

When reading/writing a field type (for a Column Definition), MySQL expects/writes a 1 byte wide [`enum_field_type`](https://dev.mysql.com/doc/dev/mysql-server/latest/field__types_8h.html#a69e798807026a0f7e12b1d6c72374854) followed by a 2 byte wide [Column Definition Flag](https://dev.mysql.com/doc/dev/mysql-server/latest/group__group__cs__column__definition__flags.html).

However, when reading a `COM_STMT_EXECUTE` payload (that specifies parameters through `new_params_bind_flag`), MySQL indicates `parameter_types` with the same 1 byte wide `enum_field_type` followed by a [1 byte wide flag that indicates signedness](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_com_stmt_execute.html#:~:text=a%20flag%20byte%20which%20has%20the%20highest%20bit%20set%20if%20the%20type%20is%20unsigned%20%5B80%5D).

So basically, read `0x80` for COM_STMT_EXECUTE parameters, but read/write `0x20` for `field_types`/column definitions.
I'm assuming MySQL does it this way because the majority of the Column Definition Flags are nonsensical/meaningless when paired up with parameters to prepared statements. Regardless, this was a subtle bug, and we should have tests for parsing `COM_STMT_EXECUTE` with `new_params_bind_flag`.

Fixes https://github.com/dolthub/dolt/issues/6728

### CODE REVIEW

File: go/sqltypes/type.go [hunk 1]
```diff
@@ -145,7 +145,7 @@ const (
 // bit-shift the mysql flags by two byte so we
 // can merge them with the mysql or vitess types.
 const (
-	mysqlUnsigned = 32
+	mysqlUnsigned = 128
```
Comment (max-hoffman) [inline go/sqltypes/type.go:148]: is it a problem that unsigned and binary are the same value?

Comment (max-hoffman) [reply go/sqltypes/type.go:148]: I'm having a hard time finding a reference for the binary flag, I guess they could be used mutually exclusively so it doesn't matter? Certainly seems easy enough for them to be different

**Comment by yx1215** [comment]
This seems to fix the issue for insert, but select still have error. See this [comment](https://github.com/dolthub/dolt/issues/6728#issuecomment-1738378874)

File: go/sqltypes/type.go (hunk 1 above)
Comment (JCOR11599) [reply go/sqltypes/type.go]: MySQL docs are difficult to look through, I updated the PR description with links to the relevant stuff.
But yeah, they shouldn't be the same value.

File: go/mysql/query.go [hunk 2]
```diff
@@ -650,7 +650,11 @@ func (c *Conn) parseComStmtExecute(prepareData map[uint32]*PrepareData, data []b
 			}
 
 			// convert MySQL type to internal type.
-			valType, err := sqltypes.MySQLToType(int64(mysqlType), int64(flags))
+			// for com_stmt_execute, the flag will either be 0x00 or 0x80 indicating signed or unsigned
+			// as a result, we need to shift the flag to the right by 2 bits
+			// while this flag may conflict with the mysqlBinary constant, it doesn't matter;
+			// the format for BLOB and TEXT parameters is the same for both binary and non-binary
+			valType, err := sqltypes.MySQLToType(int64(mysqlType), int64(flags) >> 2)
```
Comment (max-hoffman) [inline go/mysql/query.go:657]: is `0x20 >> 2` the way we convert the flag const to `0x80`?

Comment (JCOR11599) [reply go/mysql/query.go:657]: The opposite, I'm converting `0x80` to `0x20`, so the `sqltypes.MySQLToType` logic can stay the same.

Comment (max-hoffman) [reply go/mysql/query.go:657]: oh i see, yeah this all makes sense to me now

**Review by max-hoffman — APPROVED** [review]
LGTM, confusing but seems like you got to the bottom of it

**Review by zachmu — APPROVED** [review]
What's here is good and can be checked in.

But we should also have tests that start a server and talk to it with the go sql driver using prepared statements for some of the boundary conditions, like in the original bug.

Dolt has a few tests like this (`dolt_server_test.go`), and we could use the same thing in gms.

**Comment by fulghum** [comment]
@jennifersp – Seems like this might be another good case that the end-to-end (i.e. including wire serialization) test harness could catch.
